### PR TITLE
fix(puppeteer): support puppeteer v15

### PR DESCRIPTION
**Pull request by michael-siek**
See: #534

### CODE REVIEW

File: packages/puppeteer/src/legacy.ts [hunk 1]
```diff
@@ -61,7 +61,7 @@ function injectJSSource(
   source: string | Function,
   args: any[] = []
 ): Promise<void> {
-  return frame.evaluate(source as any, ...args);
+  return frame.evaluate(source as any, ...args) as Promise<void>;
```
Comment (michael-siek) [inline packages/puppeteer/src/legacy.ts:64]: Needed to typecast since this was type `Promise<unknown>`

File: packages/puppeteer/tsconfig.json [hunk 2]
```diff
@@ -18,7 +18,10 @@
     "noImplicitThis": true,
     "noImplicitReturns": true,
     "moduleResolution": "node",
-    "esModuleInterop": true
+    "esModuleInterop": true,
+    "paths": {
```
Comment (stephenmathieson) [inline packages/puppeteer/tsconfig.json:22]: Why was this necessary?

Comment (michael-siek) [reply packages/puppeteer/tsconfig.json:22]: Without this it uses the types provided by puppeteer which doesn't export `JSONArray, JSONObject, ElementHandle` but the types file does

Comment (michael-siek) [reply packages/puppeteer/tsconfig.json:22]: We use types that are not exported by puppeteer and use `@types/puppeteer`

Comment (stephenmathieson) [reply packages/puppeteer/tsconfig.json:22]: Where do the types come from? I still don't understand. Why does adding `paths` change anything? Isn't this default behavior?

Comment (straker) [reply packages/puppeteer/tsconfig.json:22]: Don't we usually fix this using the `typesRoot` property?

```
"typeRoots": ["node_modules/@types"]
```

Comment (stephenmathieson) [reply packages/puppeteer/tsconfig.json:22]: From https://github.com/puppeteer/puppeteer/pull/8547:


> [BREAKING CHANGE] We've removed the ability to declare return types in type arguments for the following reasons:
> - Setting them will indubitably break auto-deduction.
> - You can just use `as ...` in TypeScript to coerce the correct type (given it makes sense).

Why do we have to hack TS rather than casting (`as ...`)?

Comment (stephenmathieson) [reply packages/puppeteer/tsconfig.json]: That `typeRoots` value is also the default. Specifying it does nothing.

Comment (stephenmathieson) [reply packages/puppeteer/tsconfig.json]: I've solved this without any hacks: https://github.com/dequelabs/axe-core-npm/pull/535/commits/b30f8a6a6de52bb6eb5c8aa6b23696eaccf4d529

**Review by stephenmathieson — COMMENTED** [review]
I should not review this, as I've made changes here.

IMO this should show up in release notes, either as a bug fix or a feature.

We should _always_ look for alternatives to hacking `tsconfig`. By using the _actual_ Puppeteer types, we can avoid all of the hacks/casts and just pass plain objects around.

**Comment by straker** [comment]
Did some quick testing using puppeteer v12-15. All ran without issues (including with typescript).
